### PR TITLE
Add username colour override

### PR DESCRIPTION
**Pull request by mitchnull**
Implement part of issue [element-web#11816](https://github.com/vector-im/element-web/issues/11816)

- allow changing the nick color by clicking the "Override color" menu item in the room member detail page.

- the override-color can be specified as a hex string (#rrggbb) or as palette index (2)

- leaving the field blank reverts to the default hash-based nick color

- the setting is stored in account_data as im.vector.setting.override_colors, which is a map of userId -> colorSpec

The same feature is implemented in element-android in [PR#2614](https://github.com/vector-im/element-android/pull/2614)

future improvements / notes:

- replace the text-based color entry with a proper color picker dialog

Thanks to `Michaelt3chguy[m` for his help on irc

Signed-off-by: Péter Radics <mitchnull@gmail.com>

Screenshots:

Override color with color index from palette (same effect as-if the userId hash would be '4'):
override-color-with-palette-color

Override color with rgb color, default palette is not used:
override-color-with-rgb-color

The new menu item is "Override color" under the avatar.
The color input dialog should be replaced with a proper color picker that allows picking a color from the palette or an arbitrary rgb color.  Using palette colors has the advantage that it presumably works well with all themes, yet it's still possible to change a frequent contact's color if it clashes with another one (or our own).

---
This PR currently has no changelog labels, so will not be included in changelogs.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.

### CODE REVIEW

**Comment by jryans** [comment]
I believe this should start with a Product review before looking at the code.

**Comment by nadonomy** [comment]
@mitchnull please can you include screenshots to aid review? Thanks!

**Comment by mitchnull** [comment]
> @mitchnull please can you include screenshots to aid review? Thanks!

Updated the original description with screenshots.

**Comment by SimonBrandner** [comment]
May I have two suggestions? 

+ Add a placeholder so that the user which format to use
+ Maybe relabel the _Override colour_ to _Override username colour_ or to _Override nick colour_

**Comment by mitchnull** [comment]
> May I have two suggestions?
> 
> * Add a placeholder so that the user which format to use
Well, ideally this input should be replaced with a proper color picker, but yes, I can add a placeholder like '#483b8b' (although it's hard to communicate the palette index part with just the placeholder...)

> * Maybe relabel the _Override colour_ to _Override username colour_ or to _Override nick colour_
Sure, will do if the general idea is not rejected for other reasons.

**Comment by dbkr** [comment]
As a new feature, this will also need tests before we can merge it, but probably worth waiting for product review first.

**Comment by mitchnull** [comment]
FYI, the android client support has been merged via https://github.com/vector-im/element-android/pull/4826

**Review by kittykat — CHANGES_REQUESTED** [review]
@mitchnull Thank you for your PR. We would like to accept it to match implementation on Android. Here are some points:
 * The menu item in the left panel should be "Override display name colour" (this needs to be updated on Android)
 * … and should be located below "Message", above "Jump to read receipt" (also needs updating on Android)
 * The dialog should have the same title: "Override display name colour"
 * The text entry in the dialog should suggest the format , e.g. "#000000" as a hint when the text entry is focused)

**Comment by SimonBrandner** [comment]
@mitchnull, are you still working on this?

**Comment by mitchnull** [comment]
> @mitchnull, are you still working on this?

I've tried a quick rebase but it went rather ugly (no wonder after more than a year neglect...) Then I couldn't get the develop branch to build and I ran out of time... At this point it's probably easier to basically reimplement it on top of current `develop`.  I'll try to give it another go next week, but I can't promise anything.  If anyone beats me to it though, it'd be even better :)

**Comment by mitchnull** [comment]
I did the rebase and apart from the error handling updated it based on the comments.  Would be nice if someone who actually knows this stuff would take over from here :)

**Comment by kittykat** [comment]
Thanks for the update! From the Netlify build:

Screenshot from 2022-05-13 14-22-56
Screenshot from 2022-05-13 14-22-25

Text entry guidelines from our designers:
Input fields (1)

There's a PR which will be landing soon to fix how text entries are done in settings, which may be a good place to reference for this.

**Comment by turt2live** [comment]
It looks like this PR is looking for a maintainer at the moment and will need updating to our modern code standards. To avoid this being tracked as outstanding work (as it's unactionable by us), I'm going to close it for now.

This isn't to say the feature/PR is rejected, despite what GitHub will imply. We just can't take it on as a team at the moment and the community hasn't really engaged with taking it over either, so we're hoping to get it off our immediate tracking states. We'll be putting the [issue](https://github.com/vector-im/element-web/issues/11816) onto a roadmap somewhere, though it'd be much faster for a contributor to take on this change than us. 

If someone out there would like to take over the PR, please do by opening a new PR incorporating the changes here (cherrypicking across to your feature branch). Equally, if you're able to take this on @mitchnull then we can reopen this.

As mentioned, we're somewhat more in favour of a colour picker if it can work across our supported browsers. This should also eliminate the use of a dialog.

**Pull request closed without merging**
